### PR TITLE
lds: extract filter chain manager from listener

### PR DESCRIPTION
**Pull request by silentdai**
Signed-off-by: Yuchen Dai <silentdai@gmail.com>

Description: 
Preparation of FCDS. With FCDS the construction of listener and filter chains should be decoupled. 
Another motivation is to split the huge `listener_manager_impl.cc`

This PR is the first step. The new physical design allows future changes at filter chain management without touching much of the listener.

Risk Level: Med
Testing: The test of filterchainmanager is left in listener manager test. Will split in future PRs.

### CODE REVIEW

**Comment by lizan** [comment]
/retest
At a quick skim this looks good, assigning @htuch. I think this should be at least Med risk even it is just a refactor, as it is large enough in core components.

**Comment by silentdai** [comment]
@yanavlasov Absolutely. I will explore how to split and simplify the listener_manager tests. And it's nice to leave the tests for the first PR to prove that the refactor doesn't break anything :)

**Comment by silentdai** [comment]
@lizan Thanks! What else do I need to do when the PR qualifies `Med`?

**Comment by lizan** [comment]
@silentdai just mark it in PR description (which goes to commit message when a maintainer merges it).

**Comment by silentdai** [comment]
@lizan This is easy. Done.

**Comment by htuch** [comment]
I think this is basically a totally mechanical code movement and refactor. @silentdai are there any functional changes, i.e. have an of the method bodies changed? @lizan if this is just code movement, I think the risk remains low.

**Comment by lizan** [comment]
@htuch While it looks like mechanical code movement, IMO this kind of refactoring (extract part of class out of it) can affect initialization/destruction order which may cause crash which we've seen in this class before. If it is just a rename I agree it is low risk.

File: source/server/listener_manager_impl.cc [hunk 1]
```diff
@@ -201,7 +202,8 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, const std::st
       local_drain_manager_(parent.factory_.createDrainManager(config.drain_type())),
       config_(config), version_info_(version_info),
       listener_filters_timeout_(
-          PROTOBUF_GET_MS_OR_DEFAULT(config, listener_filters_timeout, 15000)) {
+          PROTOBUF_GET_MS_OR_DEFAULT(config, listener_filters_timeout, 15000)),
+      filter_chain_manager_(std::make_unique<FilterChainManagerImpl>()) {
```
Comment (silentdai) [inline source/server/listener_manager_impl.cc]: Here

File: source/server/BUILD [hunk 2]
```diff
@@ -250,6 +250,41 @@ envoy_cc_library(
...
+        "@envoy_api//envoy/admin/v2alpha:config_dump_cc",
+        "@envoy_api//envoy/api/v2:lds_cc",
+    ],
+)
+
+envoy_cc_library(
+    name = "filter_chain_manager_lib",
+    srcs = ["filter_chain_manager_impl.cc"],
+    hdrs = ["filter_chain_manager_impl.h"],
     deps = [
```
Comment (lizan) [inline source/server/BUILD]: I don't think you need all deps from listener_manager_impl, can this be less? same for above.

Comment (silentdai) [reply source/server/BUILD]: Great catch! I believe the list here is beyond what is needed. 
I will manually remove it. Is there a tool to do that?

File: source/server/listener_manager_impl.h [hunk 3]
```diff
@@ -464,27 +360,7 @@ class ListenerImpl : public Network::ListenerConfig,
...
-  }
-
-  const std::vector<Network::FilterFactoryCb>& networkFilterFactories() const override {
-    return filters_factory_;
-  }
-
-private:
-  const Network::TransportSocketFactoryPtr transport_socket_factory_;
-  const std::vector<Network::FilterFactoryCb> filters_factory_;
+  std::unique_ptr<FilterChainManagerImpl> filter_chain_manager_;
```
Comment (lizan) [inline source/server/listener_manager_impl.h]: This can be just a `FilterChainManagerImpl` instead of a unique_ptr?

Comment (silentdai) [reply source/server/listener_manager_impl.h]: At this moment, yes. Will fix along with the commit.

Comment (silentdai) [reply source/server/listener_manager_impl.h]: fixed

File: source/server/BUILD (hunk 2 above)
Comment (silentdai) [reply source/server/BUILD]: A clean up by brain. tested by build filter_chain_manager_lib and listener_manager_lib

File: source/server/filter_chain_manager_impl.h [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+namespace Server {
+
+/**
+ * Implementation of FilterChainManager.
+ */
+class FilterChainManagerImpl : public Network::FilterChainManager,
+                               Logger::Loggable<Logger::Id::config> {
+public:
+  FilterChainManagerImpl();
+  virtual ~FilterChainManagerImpl() {}
```
Comment (htuch) [inline source/server/filter_chain_manager_impl.h]: Nit: this is not needed

Comment (silentdai) [reply source/server/filter_chain_manager_impl.h]: fixed

Comment (silentdai) [reply source/server/filter_chain_manager_impl.h]: @htuch  Could you take another look?

**Pull request merged**